src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt

```diff
@@ -2,70 +2,142 @@ package endpoint.eth
     }
 
     /**
      * The method passes the query with transaction to Iroha and return the answer
      * @param request - user's request with transaction hash
      * @return Transaction which is appeared in Iroha or error
      */
-    private fun performQuery(request: EthRefundRequest): Result<iroha.protocol.BlockOuterClass.Transaction, NotaryException> {
-        val query = ModelQueryBuilder()
-        val hashes = HashVector()
-        hashes.add(Hash(request.irohaTx))
-        val signedQuery = irohaSigner.signQuery(query.getTransactions(hashes).build())
+    private fun performQuery(request: EthRefundRequest): Result<iroha.protocol.BlockOuterClass.Transaction, Exception> {
+        return Result.of {
+            val hashes = HashVector()
+            hashes.add(Hash.fromHexString(request.irohaTx))
+
+            val uquery = ModelQueryBuilder().creatorAccountId(CONFIG[ConfigKeys.irohaCreator])
+                .queryCounter(BigInteger.valueOf(1))
+                .createdTime(BigInteger.valueOf(System.currentTimeMillis()))
+                .getTransactions(hashes)
+                .build()
+            val queryBlob = ModelProtoQuery(uquery).signAndAddSignature(keypair).finish().blob().toByteArray()
+            val protoQuery = Queries.Query.parseFrom(queryBlob)
 
-        TODO("send query to iroha and return answer")
+            val irohaHost = CONFIG[ConfigKeys.irohaHostname]
+            val irohaPort = CONFIG[ConfigKeys.irohaPort]
+            val channel = ManagedChannelBuilder.forAddress(irohaHost, irohaPort).usePlaintext(true).build()
+            val queryStub = QueryServiceGrpc.newBlockingStub(channel)
+            val queryResponse = queryStub.find(protoQuery)
 
+            val fieldDescriptor =
+                queryResponse.descriptorForType.findFieldByName("transactions_response")
+            if (!queryResponse.hasField(fieldDescriptor)) {
+                throw NotaryException("Query response error ${queryResponse.errorResponse}")
+            }
+
+            // return transaction
+            queryResponse.transactionsResponse.transactionsList[0]
+        }
     }
 
     /**
      * The method checks transaction and create refund if it is correct
      * @param appearedTx - target transaction from Iroha
+     * @param request - user's request with transaction hash
      * @return Refund or error
      */
-    private fun checkTransaction(appearedTx: iroha.protocol.BlockOuterClass.Transaction): Result<EthRefund, NotaryException> {
-        val commands = appearedTx.payload.getCommands(0)
-        // rollback case
-        if (commands.hasSetAccountDetail()) {
-            TODO("rollback case")
-        }
+    private fun checkTransaction(
+        appearedTx: iroha.protocol.BlockOuterClass.Transaction,
+        request: EthRefundRequest
+    ): Result<EthRefund, Exception> {
+        return Result.of {
+            val commands = appearedTx.payload.getCommands(0)
```

Comment:
You can use Commands.fromProto()

Reply:
I added task to Jira to move proto out of logic.